util/devel/check_paths

```diff
@@ -0,0 +1,176 @@
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Try to identify references to non-existent files, e.g. doc/foo
+instead of doc/rst/foo.  This script recognizes potential file
+references by recognizing certain strings as likely to be the prefix
+of a path in the Chapel repository, e.g., "doc/" or "modules/" or
+"test/".
+"""
+
+import os.path
+import argparse
+
+def check_path(p):
+    """returns True if p exists, False if it doesn't"""
+
+    # These are common written-out alternatives, not paths
+    if p == "compiler/runtime":
+        return True
+    if p == "compiler/module":
+        return True
+
+    # These show up in generated html files relative to a "modules"
+    # directory that isn't the top-level one we're going to look in.
+    if p == "modules/chpldoc.doc.html":
+        return True
+    if p == "modules/chpldoc.doc/Defined.html":
+        return True
+    # Or in generated .js
+    if p == "modules/chpldoc.doc" or p == "modules/chpldoc.doc/Defined":
+        return True
+    # jquery .js file means a different test/ directory too
+    if p == "test/unit/core.js":
+        return True
+
+    return os.path.isfile(p) or os.path.isdir(p)
+
+# Memoize will be more useful once the script runs on multiple files
+# in one go.
+# Cribbed from a stackoverflow answer
+class Memoize:
+    def __init__(self, f):
+        self.f = f
+        self.memo = {}
+    def __call__(self, *args):
+        if not args in self.memo:
+            self.memo[args] = self.f(*args)
+        return self.memo[args]
+
+@Memoize
+def check_token(t):
+    """If t "looks like" a path but doesn't exist, return False.
+    Otherwise, return True."""
+
+    # Skip everything that doesn't look like something
+    if  not (t.startswith("doc/") or
+             t.startswith("test/") or
+             t.startswith("compiler/") or
+             t.startswith("runtime/") or
+             t.startswith("modules/") or
+             t.startswith("spec/") or
+             t.startswith("util/") or
+             t.startswith("tools/") or
+             t.startswith("third-party/") or
+             t.startswith("CHPL_HOME/") or
+             t.startswith("//github.com/chapel-lang/chapel/blob/master/")):
+        return True
+
+    elif t.startswith("CHPL_HOME/"):
+        t = t[10:]
+    elif t.startswith("//github.com/chapel-lang/chapel/blob/master/"):
+        t = t[44:]
+
+    # People like to say: "For details, see the README...".
+    while t.endswith("."):
+        t = t[:-1]
+
+    # Don't worry if it was "CHPL_HOME/..."
```

Comment:
I wonder if these types of exceptions should also be encapsulated into a separate function

Reply:
Oh, sorry, I misread this comment before and thought you were suggesting a separate file.

But I'm not sure about splitting the function up.  Everything in this ``check_token()`` function seems pretty equally to be part of checking this one token.  I'll sleep on it.